Initial support for ARM64 Windows

cocoa-xu opened the pull request:
Hi, this PR includes initial patches and instructions to bring Erlang/OTP to ARM64 Windows. These changes should only affect the behaviour when running in WSL and user executes one of the following command to set up the build environment

- `./otp_build env_win32 arm64`, set up build environment for compiling ARM64 Erlang/OTP natively,
- `./otp_build env_win32 x64_arm64`, set up build environment for cross-compiling ARM64 Erlang/OTP on an x86_64 machine

Although I understand that ARM64 Windows might not be anywhere close to the main tier of support yet, this PR can serve as an initial guide for people in the community who're also interested in running Erlang/OTP natively on ARM64 Windows.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=8142) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=8142) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=8142) it.</sub>

cocoa-xu commented on erts/etc/win32/wsl_tools/reg_query.sh:
```diff
@@ -10,6 +10,8 @@ BACKED=`echo "$1" | sed 's,/,\\\\,g'`
 
 if [ $CONFIG_SUBTYPE = "win64" ]; then
     REG_OPT=" /reg:64"
+elif [ X"$CONFIG_SUBTYPE" = X"arm64" -o X"$CONFIG_SUBTYPE" = X"x86_arm64" ]; then
```
Sorry this was a typo, I'll fix it soon!

cocoa-xu commented on erts/etc/win32/wsl_tools/vc/emu_cc.sh:
```diff
@@ -23,6 +23,10 @@
+elif [ X"$CONFIG_SUBTYPE" = X"arm64" ]; then
+    GCC="aarch64-w64-mingw32-gcc -m64"
+elif [ X"$CONFIG_SUBTYPE" = X"x64_arm64" ]; then
+    GCC="aarch64-w64-mingw32-clang -m64"
```
Hi @jhogberg, `x64_arm64` is used when cross-compiling to arm64 windows on an x86_64 machine because we need to call `vcvarsall.bat` in `erts/etc/win32/wsl_tools/SetupWSLcross.bat` with different parameters to get right msvc toolchain, for example,

- call `vcvarsall.bat arm64` if compile natively
- call `vcvarsall.bat amd64_arm64` if cross-compile on x86_64 machines

And I think that it's somewhat awkward naming on Windows because `win32` and `win64` traditionally only stands for 32-bit and 64-bit respectively, but they actually means `x86` and `x86_64`. I wonder if we should standardise these naming while providing "aliases" so that build script won't break. I'm think that adding a universal entry point for setting up environment variables:

`./otp_build env_win [<BUILD_ARCH>-]<TARGET_ARCH>`

where `BUILD_ARCH` and `TARGET_ARCH` each can be one of `x86`, `x86_64` and `aarch64`.

For example,

### x86 target
For compiling natively on x86 windows, `env_win32` should be the alias of `env_win x86` (or vice-versa). The following commands should behave the same

- `./otp_build env_win x86`, universal entry point
- `./otp_build env_win32`, alias for backward compatibility

For cross-compiling, we will have 
- `./otp_build env_win x86_64-x86`, universal entry point
- `./otp_build env_win aarch64-x86`, universal entry point

### x86_64 target
Similarly, the following commands should behave the same

- `./otp_build env_win x86_64`, universal entry point
- `./otp_build env_win32 x64`, alias for backward compatibility
- `./otp_build env_win32 x64`, alias for backward compatibility
- `./otp_build env_win64 x64`, alias for backward compatibility

For cross-compiling, we will have 
- `./otp_build env_win x86-x86_64`, universal entry point
- `./otp_build env_win aarch64-x86_64`, universal entry point

### aarch64 target
For native compiling,
- `./otp_build env_win aarch64`, universal entry point

Maybe we can have these aliases which would follow the convention:
- `./otp_build env_win32 x86-aarch64`
- `./otp_build env_win32 x64-aarch64`
- `./otp_build env_win64 x86-aarch64`
- `./otp_build env_win64 x64-aarch64`

And for cross-compiling, we will have 
- `./otp_build env_win x86-aarch64 `, universal entry point
- `./otp_build env_win x86_64-aarch64`, universal entry point

WDYT?

cocoa-xu replied:
The reason why using `clang` for cross-compiling is that `aarch64-w64-mingw32-gcc` seems to crash the windows machine😅 I'll double verify this on another x86_64 windows machine.

cocoa-xu replied:
> What GCC version is `aarch64-w64-mingw32-gcc`? There's a known bug in GCC 7 (maybe 8 or 9 too?) where it segfaults while compiling the JIT. If so, I don't think we want this workaround.

It's actually LLVM with a wrapper script, https://github.com/mstorsjo/llvm-mingw/releases. I tested `aarch64-w64-mingw32-gcc` on a physical x86_64 machine, and it seems to be okay now (as I'm not able to do a full build -- there're still some issues when cross-compiling). 

I think the reason was that I was testing on a virtual machine, and perhaps there were some bugs in the virtualization layer/software.

cocoa-xu commented on erts/emulator/sys/win32/erl_win_sys.h:
```diff
@@ -192,6 +192,9 @@ struct erts_sys_time_read_only_data__ {
     ErtsMonotonicTime (*os_monotonic_time)(void);
     void (*os_times)(ErtsMonotonicTime *, ErtsSystemTime*);
     ErtsSysHrTime (*sys_hrtime)(void);
+    ErtsSysPerfCounter (*perf_counter)(void);
+    ErtsSysPerfCounter perf_counter_unit;
+    int ticks_per_sec;
```
Will do!

cocoa-xu commented on erts/etc/win32/wsl_tools/vc/cc.sh:
```diff
@@ -29,6 +29,14 @@ SAVE="$@"
 
+if [ "$CONFIG_SUBTYPE" = "x64_arm64" -o "$CONFIG_SUBTYPE" = "arm64" ]; then
+	MACHINE="ARM64"
+	STDCXX_VERSION="/std:c++20"
```
I think there're some one-offs in the code base, or at least LLVM think they should be in C++20 standard. For example,

```
CXX    obj/win32/opt/jit/beam_asm_global.o
beam\jit\arm\beam_asm_global.cpp(77): error C7555: use of designated initializers requires at least '/std:c++20'
Failed: cl.exe -nologo -D__WIN32__ -DWIN32 -DWINDOWS -D_WIN32 -DNT -D_CRT_SECURE_NO_DEPRECATE -D__aarch64__ -MD -Ox -Z7 /std:c++17 /Zc:__cplusplus -DASMJIT_EMBED=1 -DASMJIT_NO_BUILDER=1 -DASMJIT_NO_DEPRECATED=1 -DASMJIT_STATIC=1 -DASMJIT_NO_FOREIGN=1 -Iwin32/opt/jit -Ibeam -Isys/win32 -Isys/common -Iwin32 -Izlib -Ipcre -Iryu -Iopenssl/include -I../include -I../include/win32 -I../include/internal -I../include/internal/win32 -Ibeam/jit -Ibeam/jit/arm -I"C:\src\erts\win32" -D_WIN32_WINNT=0x0600 -DWINVER=0x0600 -DERTS_MIXED_VC -DSTATIC_ERLANG_DRIVER -DHAVE_CONFIG_H -DUSE_THREADS -DWIN32_THREADS -DBEAMASM=1 /std:c++17 /Zc:__cplusplus /EHsc /FS -c -FoC:\src\erts\emulator\obj\win32\opt\jit\beam_asm_global.o beam\jit\arm\beam_asm_global.cpp
make[4]: *** [win32/Makefile:947: obj/win32/opt/jit/beam_asm_global.o] Error 2
make[4]: Leaving directory '/mnt/c/src/erts/emulator'
make[3]: *** [/mnt/c/src/make/run_make.mk:35: opt] Error 2
make[3]: Leaving directory '/mnt/c/src/erts/emulator'
make[2]: *** [Makefile:45: opt] Error 2
make[2]: Leaving directory '/mnt/c/src/erts'
make[1]: *** [Makefile:60: jit] Error 2
make[1]: Leaving directory '/mnt/c/src/erts'
make: *** [Makefile:483: emulator] Error 2
```
 I'll try to fix these one-offs and keep the minimum version to be C++17.

cocoa-xu replied:
### update
Designated initializers seem to be indeed available since C++20, [reference link](https://en.cppreference.com/w/cpp/language/aggregate_initialization).

cocoa-xu replied:
Fixed in [7e9801f](https://github.com/erlang/otp/pull/8142/commits/7e9801fdb230e236033f89128e990b5925d0f03d)

dgud commented on otp_build:
```diff
@@ -1017,7 +1041,7 @@ do_debuginfo_win32 ()
     BINDIR="$ERL_TOP/bin/$TARGET"
     EVSN=`grep '^VSN' erts/vsn.mk | sed 's,^VSN.*=[^0-9]*\([0-9].*\)$,@\1,g;s,^[^@].*,,g;s,^@,,g'`
-    for f in beam.debug.smp.dll beam.smp.pdb beam.debug.smp.dll.pdb erl.pdb erlexec.pdb; do
+    for f in beam.debug.smp.dll beam.smp.pdb beam.debug.smp.dll.pdb werl.pdb erl.pdb erlexec.pdb; do
```
I guess werl.pdb is not needed anymore since werl is just a copy of erl?

cocoa-xu replied:
Okay, I'll remove it!

cocoa-xu commented on make/autoconf/otp.m4:
```diff
@@ -35,37 +35,79 @@ AC_DEFUN([ERL_CANONICAL_SYSTEM_TYPE],
     # Adjust for local legacy windows hack...
     AS_CASE([$host],
-            [local-*-windows],
+            [local-aarch64-*-winodws],
```
Ahhh! Sorry for the typo, I'll fix it now!

cocoa-xu commented on lib/common_test/configure:
```diff
@@ -2044,6 +2044,13 @@ test -n "$target_alias" &&
+                host_os=win32
+                host_vendor=
+                host_cpu=aarch64
+             ;; #(
```
I compiled autoconf 2.71, and updated the configure files.  It should be okay now.

cocoa-xu commented:
### Some updates

wxWidget also seems to be okay now.

![ARM64-OTP-4](https://github.com/erlang/otp/assets/89497197/27f19bb8-3902-45ab-b3b6-879111d51486)
![ARM64-OTP-5](https://github.com/erlang/otp/assets/89497197/8c345528-44bd-47fb-8869-94d6aba380d5)

As for the nsis script, I'm not sure what's the ARM64's equivalent of `$PROGRAMFILES64` so that I can get it to install Erlang/OTP in `C:\Program Files (Arm)` by default.

reflectronic commented:
> so that I can get it to install Erlang/OTP in C:\Program Files (Arm) by default.

`Program Files (Arm)` is for Arm32 applications; 64-bit applications should always be installed in `Program Files`.

ref: https://learn.microsoft.com/en-us/windows/arm/faq#what-program-files-folder-should-i-use-to-install-my-arm64-application-

cocoa-xu commented:
> > so that I can get it to install Erlang/OTP in C:\Program Files (Arm) by default.
> 
> `Program Files (Arm)` is for Arm32 applications; 64-bit applications should always be installed in `Program Files`.
> 
> ref: https://learn.microsoft.com/en-us/windows/arm/faq#what-program-files-folder-should-i-use-to-install-my-arm64-application-

Oh I didn't know that before. Thanks for the information! Gonna update the nsis script file!

dgud commented on erts/etc/win32/nsis/erlang20.nsi:
```diff
@@ -68,7 +68,11 @@ Var STARTMENU_FOLDER
+	!if ${WINTYPE} == "arm64"
+		!define MUI_STARTMENUPAGE_DEFAULTFOLDER "${OTP_PRODUCT} ${OTP_RELEASE} (arm64)"
+	!else
+		!define MUI_STARTMENUPAGE_DEFAULTFOLDER "${OTP_PRODUCT} ${OTP_RELEASE} (x64)"
+	!endif
```
Was (i386) ?

cocoa-xu replied:
Sorry for the copy-paste error. It's fixed now

dgud commented on erts/etc/win32/nsis/find_redist.sh:
```diff
@@ -101,11 +107,16 @@ if [ x"$VCToolsRedistDir" != x"" ]; then
 CLPATH=`lookup_prog_in_path cl`
 if [ -z "$CLPATH" ]; then 
-    echo "Can not locate cl.exe and vcredist_x86/x64.exe - OK if using mingw" >&2
+    echo "Can not locate cl.exe and vcredist_x86/x64/arm64.exe - OK if using mingw" >&2
```
```suggestion
    echo "Can not locate cl.exe and thus not vcredist - OK if using mingw" >&2
```

dgud replied:
Something like this?

cocoa-xu commented:
And here we have the test results. I'm not sure which of these files should I archive and upload here (and also because the test result directory is about ~1.5 GB, and 259MB after compressing), so I only did a screenshot for now. Please let me know if you need me to upload the test results.

![IMG_6891CE8947DB-1](https://github.com/erlang/otp/assets/89497197/0859f782-a722-4059-b321-fc9fe964c5f4)

cocoa-xu commented:
> It seems that the `emulator` and `kernel` tests crashed hard, can you see which tests it crashed on?

Hi, I decided to upload the test result here, https://github.com/cocoa-xu/otp-build/releases/download/v27.0-rc1/test_server-otp_win_arm64_27.0-rc1.tar.xz, and using xz reduces its size to 75MB. It will be easier to check all relevant logs, as well as for other people to look and maybe give any feedbacks.

It corresponds to this installer, https://github.com/cocoa-xu/otp-build/releases/download/v27.0-rc1/otp_win_arm64_27.0-rc1.exe.

cocoa-xu commented:
> It seems that the `emulator` and `kernel` tests crashed hard, can you see which tests it crashed on?

As for the ones that cause crashes, they're `bs_construct_SUITE:error_info/1` and `logger_disk_log_h_SUITE:disk_log_full/1`

#### `bs_construct_SUITE:error_info/1`

<details>

```
=== Test case: [bs_construct_SUITE:error_info/1](file:///Users/cocoa/Downloads/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.40.17/tests.emulator_test.logs/run.2024-02-20_06.40.46/bs_construct_suite.src.html#error_info-1) (click for source code)

=== Config value:

    [{watchdog,<0.115844.16>},
     {tc_logfile,"c:/src/release/tests/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.40.17/tests.emulator_test.logs/run.2024-02-20_06.40.46/bs_construct_suite.error_info.html"},
     {tc_group_properties,[]},
     {tc_group_path,[]},
     {data_dir,"c:/src/release/tests/emulator_test/bs_construct_SUITE_data/"},
     {priv_dir,"c:/src/release/tests/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.40.17/tests.emulator_test.logs/run.2024-02-20_06.40.46/log_private/"},
     {nodenames,[]}]

=== Current directory is "c:/src/release/tests/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.40.17"

=== Started at 2024-02-20 06:45:21



<< Atom / binary , Binary / binary >>: segment 1 of type 'binary': expected a binary but got: some_atom

<< Binary / binary , Atom / binary >>: segment 2 of type 'binary': expected a binary but got: some_atom

<< 1 : 32 , Binary / binary , Atom / binary >>: segment 3 of type 'binary': expected a binary but got: some_atom

<< 1 : 32 , "xyz" , Binary / binary , Atom / binary >>: segment 4 of type 'binary': expected a binary but got: some_atom

<< Atom : 32 >>: segment 1 of type 'integer': expected an integer but got: some_atom

<< Atom : ( id ( 32 ) ) >>: segment 1 of type 'integer': expected an integer but got: some_atom

<< LongList : 32 >>: segment 1 of type 'integer': expected an integer but got: [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19|...]

<< 42 : Atom >>: segment 1 of type 'integer': expected a non-negative integer as size but got: some_atom

<< Atom : 32 >>: segment 1 of type 'integer': expected an integer but got: some_atom

<< 42 : NegSize >>: segment 1 of type 'integer': expected a non-negative integer as size but got: -1

<< 42 : HugeNegSize >>: segment 1 of type 'integer': expected a non-negative integer as size but got: -18446744073709551616

<< 42 : ( 1 bsl 58 ) / unit : 255 >>: segment 1 of type 'integer': the size 288230376151711744 is too large

<< 42 : ( 1 bsl 60 ) / unit : 8 >>: segment 1 of type 'integer': the size 1152921504606846976 is too large

<< 42 : ( 1 bsl 64 ) >>: segment 1 of type 'integer': the size 18446744073709551616 is too large

<< Atom : 10 / binary >>: segment 1 of type 'binary': expected a binary but got: some_atom

<< Atom : ( id ( 10 ) ) / binary >>: segment 1 of type 'binary': expected a binary but got: some_atom

<< Binary : Atom / binary >>: segment 1 of type 'binary': expected a non-negative integer as size but got: some_atom

<< Binary : NegSize / binary >>: segment 1 of type 'binary': expected a non-negative integer as size but got: -1

<< Binary : HugeNegSize / binary >>: segment 1 of type 'binary': expected a non-negative integer as size but got: -18446744073709551616

<< Binary : BadSize / binary >>: segment 1 of type 'binary': expected a non-negative integer as size but got: bad_size

<< BadBinary : BadSize / binary >>: segment 1 of type 'binary': expected a non-negative integer as size but got: bad_size

<< Binary : 10 / binary >>: segment 1 of type 'binary': the value <<"abc">> is shorter than the size of the segment

<< Binary : ( id ( 10 ) ) / binary >>: segment 1 of type 'binary': the value <<"abc">> is shorter than the size of the segment

<< Atom / binary >>: segment 1 of type 'binary': expected a binary but got: some_atom

<< ( id ( << 1 : 1 >> ) ) / binary >>: segment 1 of type 'binary': the size of the value <<1:1>> is not a multiple of the unit for the segment

<< ( id ( << 0 : 1111 >> ) ) / binary >>: segment 1 of type 'binary': the size of the value <<0,0,0,0,0,0,0,0...,0,0:7>> is not a multiple of the unit for the segment

<< 0 , ( id ( << 1 : 1 >> ) ) / binary >>: segment 2 of type 'binary': the size of the value <<1:1>> is not a multiple of the unit for the segment

<< 0 , ( id ( << 0 : 1111 >> ) ) / binary >>: segment 2 of type 'binary': the size of the value <<0,0,0,0,0,0,0,0...,0,0:7>> is not a multiple of the unit for the segment

<< Binary : ( 1 bsl 64 ) / binary >>: segment 1 of type 'binary': the size 18446744073709551616 is too large

<< Binary : ( id ( 1 bsl 64 ) ) / binary >>: segment 1 of type 'binary': the size 18446744073709551616 is too large

<< Atom : 64 / float >>: segment 1 of type 'float': expected a float or an integer but got: some_atom

<< Atom : Atom / float >>: segment 1 of type 'float': expected a non-negative integer as size but got: some_atom

<< 42.0 : NegSize / float >>: segment 1 of type 'float': expected a non-negative integer as size but got: -1

<< 42.0 : HugeNegSize / float >>: segment 1 of type 'float': expected a non-negative integer as size but got: -18446744073709551616

<< 42.0 : ( id ( 1 ) ) / float >>: segment 1 of type 'float': expected one of the supported sizes 16, 32, or 64 but got: 1

<< HugeBig : ( id ( 64 ) ) / float >>: segment 1 of type 'float': the value 350746621104...713208549376 is outside the range expressible as a float

<< HugeBig : 64 / float >>: segment 1 of type 'float': the value 350746621104...713208549376 is outside the range expressible as a float

<< 42.0 : ( id ( 1 bsl 64 ) ) / float >>: segment 1 of type 'float': the size 18446744073709551616 is too large

<< Atom / utf8 >>: segment 1 of type 'utf8': expected a non-negative integer encodable as utf8 but got: some_atom

<< Atom / utf16 >>: segment 1 of type 'utf16': expected a non-negative integer encodable as utf16 but got: some_atom

<< Atom / utf32 >>: segment 1 of type 'utf32': expected a non-negative integer encodable as utf32 but got: some_atom

<< 0 : ( MaxSmall ) / unit : 32 , 0 : 64 >>: segment 1 of type 'integer': the size 576460752303423487 is too large

<< 0 : ( MaxSmall ) / unit : 32 , ( id ( 0 ) ) : 64 >>: segment 1 of type 'integer': the size 576460752303423487 is too large
```

</details>

#### `logger_disk_log_h_SUITE:disk_log_full/1`

<details>

```
*** System report during logger_disk_log_h_SUITE:disk_log_full/1 in tty 2024-02-20 07:35:45.575 ***[🔗](file:///Users/cocoa/Downloads/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.45.23/tests.kernel_test.logs/run.2024-02-20_06.45.55/logger_disk_log_h_suite.disk_log_full.html#e-1363)
=ERROR REPORT==== 20-Feb-2024::07:35:45.575000 ===
** Generic server <0.106479.0> terminating 
** Last message in was {disk_log,'test_server@COCOA-ARM64-PC',
                                 logger_disk_log_h_SUITE,full}
** When Server state == #{id => logger_disk_log_h_logger_disk_log_h_SUITE,
                          module => logger_h_common,mode => async,
                          idle => false,burst_limit_enable => true,
                          drop_mode_qlen => 200,flush_qlen => 1000,
                          sync_mode_qlen => 10,burst_msg_count => 9,
                          burst_win_ts => -24019194990094072,
                          cb_state =>
                              #{id => logger_disk_log_h_SUITE,
                                module => logger_disk_log_h,
                                filesync_repeat_interval => 5000,
                                ctrl_sync_count => 11,
                                handler_state =>
                                    #{log_opts =>
                                          #{type => halt,
                                            file =>
                                                "c:/src/release/tests/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-20_06.45.23/tests.kernel_test.logs/run.2024-02-20_06.45.55/log_private/disk_log_full",
                                            max_no_bytes => 50,
                                            max_no_files => 1},
                                      prev_log_result => ok,
                                      prev_sync_result => ok,
                                      prev_disk_log_info => undefined},
                                last_op => write,
                                rep_sync_tref =>
                                    #Ref<0.3961621595.2628255746.191094>},
                          last_load_ts => -24019194990093602,last_qlen => 0,
                          mode_ref =>
                              {logger_olp,
                                  logger_disk_log_h_logger_disk_log_h_SUITE},
                          overload_kill_restart_after => 5000,
                          burst_limit_max_count => 500,
                          burst_limit_window_time => 1000,
                          overload_kill_enable => false,
                          overload_kill_mem_size => 3000000,
                          overload_kill_qlen => 20000}
** Reason for termination ==
** {badarg,[{erlang,display_string,
                    ["Logger - error: "],
                    [{file,"erlang.erl"},
                     {line,2139},
                     {error_info,#{cause => 'The handle is invalid.\r\n',
                                   module => erl_erts_errors}}]},
            {logger,internal_log,2,[{file,"logger.erl"},{line,726}]},
            {logger_disk_log_h,error_notify_new,4,
                               [{file,"logger_disk_log_h.erl"},{line,365}]},
            {logger_h_common,handle_info,2,
                             [{file,"logger_h_common.erl"},{line,310}]},
            {logger_olp,try_callback_call,4,
                        [{file,"logger_olp.erl"},{line,617}]},
            {logger_olp,handle_info,2,[{file,"logger_olp.erl"},{line,271}]},
            {gen_server,try_handle_info,3,
                        [{file,"gen_server.erl"},{line,2158}]},
            {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,2246}]}]}
```

</details>

cocoa-xu commented:
> It seems that the `emulator` and `kernel` tests crashed hard, can you see which tests it crashed on?

Updates for the `emulator` tests. I looked and removed the lines that cause hard-crashes, and now the test results is

```
2316 Ok, 12 Failed, 208 Skipped of 2536
```

### hard-crashes
#### common
In

- `bs_construct_SUITE:error_info`
- `bs_construct_no_opt_SUITE:error_info`
- `bs_construct_r25_SUITE:error_info`
- `bs_construct_stripped_types_SUITE:error_info`

```erlang
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(id(MaxSmall))/unit:32,0:64>>),
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(id(MaxSmall))/unit:32,(id(0)):64>>),
```

#### `bs_construct_no_opt_SUITE:error_info`

```erlang
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(MaxSmall)/unit:32,0:64>>),
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(MaxSmall)/unit:32,(id(0)):64>>),
```

### failed
- `bif_SUITE:display_string`
- `bs_construct_SUITE:dynamic`
- `bs_construct_SUITE:error_info`
- `bs_construct_no_opt_SUITE:error_info`
- `bs_construct_r25_SUITE:dynamic`
- `bs_construct_r25_SUITE:error_info`
- `bs_construct_stripped_types_SUITE:error_info`
- `hello_SUITE:hello`
- `port_SUITE: pipe_limit_env`
- `port_SUITE: unregister_name`
- `smoke_test_SUITE:jump_table`
- `time_SUITE:time_unit_conversion`

### failed to init
- **`mtx_SUITE:init_per_suite`**
- **`nif_SUITE init_per_suite`**

cocoa-xu commented:
Updates for all tests.

I did a clean compile and removed the lines that cause hard-crashes in `emulator` and `kernel`.

For the `emulator` tests, the lines I removed were:

```erlang
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(id(MaxSmall))/unit:32,0:64>>),
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(id(MaxSmall))/unit:32,(id(0)):64>>),
```
In these test cases:

- `bs_construct_SUITE:error_info`
- `bs_construct_no_opt_SUITE:error_info`
- `bs_construct_r25_SUITE:error_info`
- `bs_construct_stripped_types_SUITE:error_info`

In addition to the lines above, I also have to remove these two lines in `bs_construct_no_opt_SUITE:error_info`

```erlang
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(MaxSmall)/unit:32,0:64>>),
{system_limit, {1,integer,size,MaxSmall}, _} = ?ERROR_INFO(<<0:(MaxSmall)/unit:32,(id(0)):64>>),
```

For the `kernel` tests, I need to change these two tests to

```erlang
disk_log_full(Config) ->
    ok.
% ...
disk_log_events(_Config) ->
    ok.
```

After the above changes, I ran the tests in WSL (which was done in Powershell before), and the latest test results are:

![test_results](https://github.com/erlang/otp/assets/89497197/79c8b5f0-30bf-4641-bb77-f5357ab22014)

Also, sorry that I'm not sure how I should properly remove/truncate successful tests to save space, so I still have to upload the full test results a GitHub release file [here](https://github.com/cocoa-xu/otp-build/releases/download/v27.0-rc1/test_results-otp_win_arm64_27.0-rc1.tar.xz), as the attachment file cannot exceed 25MBs in PR comments.

cocoa-xu commented:
And I was wondering if and which failed test cases should we address in this PR? Perhaps the ones in the `emulator`, `kernel` and `system`?

cocoa-xu commented on system/doc/docs.exs:
```diff
@@ -6,6 +6,7 @@
     "installation_guide/INSTALL.md": [source: "../../HOWTO/INSTALL.md"],
     "installation_guide/INSTALL-CROSS.md": [source: "../../HOWTO/INSTALL-CROSS.md"],
     "installation_guide/INSTALL-WIN32.md": [source: "../../HOWTO/INSTALL-WIN32.md"],
+    "installation_guide/INSTALL-WIN32-ARM64.md": [source: "../../HOWTO/INSTALL-WIN32-ARM64.md"],
```
This line was causing issue when testing documentation in CI (although the file should be there?), should I remove this line for now?

cocoa-xu commented:
I found that a few of these hard-crashed ones were related to `erlang:display_string/{1,2}` errored with `bad argument`. However, the call was unsuccessful because the `fd` passed to `WriteFile` is an invalid handle according to `GetLastError`.

```c
#ifdef __WIN32__
    printf("[debug] we're in display_string_2 win32\r\n");
    HANDLE fd;
    if (ERTS_IS_ATOM_STR("stdout", BIF_ARG_1)) {
        fd = GetStdHandle(STD_OUTPUT_HANDLE);
        printf("[debug] and the BIF_ARG_1 is stdout, fd=%d\r\n", fd);
    } else if (ERTS_IS_ATOM_STR("stderr", BIF_ARG_1)) {
        fd = GetStdHandle(STD_ERROR_HANDLE);
        printf("[debug] and the BIF_ARG_1 is stderr, fd=%d\r\n", fd);
        if (fd == INVALID_HANDLE_VALUE) {
            printf("[debug] and fd == INVALID_HANDLE_VALUE\r\n");
        }
    }
#else
// ...
#endif

#ifdef __WIN32__
        Uint32 w;
        if (!WriteFile(fd, str, len, &w, NULL)) {
            DWORD last_error = GetLastError();
            LPTSTR error_text = NULL;
            FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, HRESULT_FROM_WIN32(last_error), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPTSTR)&error_text, 0, NULL);
            if (error_text != NULL) {
                printf("[debug] error msg: %s\r\n", error_text);
                LocalFree(error_text);
            } else {
                printf("[debug] error msg: cannot get formatted error message\r\n");
            }
            goto error;
        }
        written = (Sint)w;
#else
// ...
#endif
```

The coresponding test case was changed to:
```erlang
display_string(Config) when is_list(Config) ->
    # using `erlang:display` for comparison and check that we can display to console
    true = erlang:display("hej"),
    true = erlang:display_string("hej"),
    ok.
```

And I got this output when running this test case:

```
$ /mnt/c/src/release/aarch64-windows-msvc/bin/erl.exe -noinput -eval "ts:run(emulator, bif_SUITE, display_string, [batch]),erlang:halt()"
Command: erl -sname test_server -rsh ssh -env PATH "c:\src\release\tests\emulator_test;c:\src\release\AARCH6~1\ERTS-1~1.2\bin;c:\Program Files (Arm)\MICROS~1\JDK-16~1.7-H\bin;c:\PROGRA~1\PARALL~1\PARALL~1\APPLIC~1;c:\Windows\System32;c:\Windows;c:\Windows\System32\wbem;c:\Windows\System32\WindowsPowerShell\v1.0;c:\Windows\System32\OpenSSH;c:\PROGRA~1\Git\cmd;c:\PROGRA~3\CHOCOL~1\bin;c:\Users\cocoa\AppData\Local\Programs\Python\PYTHON~1\Scripts;c:\Users\cocoa\AppData\Local\Programs\Python\PYTHON~1;c:\Users\cocoa\AppData\Local\MICROS~1\WINDOW~1;c:\Users\cocoa\AppData\Local\Programs\MICROS~1\bin;c:\PERL_A~1\bin;c:\Users\cocoa\AppData\Local\GITHUB~1\bin;c:\ERLANG~2\bin;c:\Users\cocoa\ELIXIR~1.1\bin;c:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1439~1.335\bin\HOSTAR~1\arm64;c:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\arm64;c:\Program Files (Arm)\MICROS~1\JDK-16~1.7-H\bin;c:\PROGRA~1\PARALL~1\PARALL~1\APPLIC~1;c:\Windows\System32;c:\Windows;c:\Windows\System32\wbem;c:\Windows\System32\WindowsPowerShell\v1.0;c:\Windows\System32\OpenSSH;c:\PROGRA~1\Git\cmd;c:\PROGRA~3\CHOCOL~1\bin;c:\Users\cocoa\AppData\Local\Programs\Python\PYTHON~1\Scripts;c:\Users\cocoa\AppData\Local\Programs\Python\PYTHON~1;c:\Users\cocoa\AppData\Local\MICROS~1\WINDOW~1;c:\Users\cocoa\AppData\Local\Programs\MICROS~1\bin;c:\PERL_A~1\bin;c:\Users\cocoa\AppData\Local\GITHUB~1\bin;c:\ERLANG~2\bin;c:\Users\cocoa\ELIXIR~1.1\bin" -env ERL_CRASH_DUMP c:/src/release/tests/test_server/emulator_erl_crash.dump -boot start_sasl -sasl errlog_type error -pz "c:/src/release/tests/test_server" -enable-feature maybe_expr -ct_test_vars "{net_dir,\"\"}" -eval "ts_run:ct_run_test(\"c:/src/release/tests/emulator_test\", [{abort_if_missing_suites,true},{suite,bif_SUITE},{testcase,display_string},{logdir,\"../test_server\"},{config,[\"../test_server/ts.config\",\"../test_server/ts.win32.config\"]},{vars,[{verbose,0}]},batch,{scale_timetraps,true}])"  -noinput -eval "erlang:halt(0,[{flush,false}])."

Common Test starting (cwd is c:/src/release/tests/emulator_test)


Common Test: Running make in test directories...
Recompile: bif_SUITE
Recompile: gen_tcp_dist
Recompile: ignore_cores

CWD set to: "c:/src/release/tests/test_server/ct_run.test_server@COCOA-ARM64-PC.2024-02-21_22.31.26"

Testing tests.emulator_test.bif_SUITE.display_string: Starting test, 1 test cases
"hej"
[debug] we're in display_string_2 win32
[debug] and the BIF_ARG_1 is stderr, fd=96
[debug] WriteFile went wrong
[debug] error msg: The handle is invalid.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
erlang:display_string failed on line 2139
Reason: {badarg,[{erlang,...},{...}|...]}
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Testing tests.emulator_test.bif_SUITE.display_string: *** FAILED test case 1 of 1 ***
Testing tests.emulator_test.bif_SUITE.display_string: TEST COMPLETE, 0 ok, 1 failed of 1 test cases

Updating c:/src/release/tests/test_server/index.html ... done
Updating c:/src/release/tests/test_server/all_runs.html ... done
```

And I also attempted to use `_fileno` and `_write` instead of `GetStdHandle` and `WriteFile` but it ended up pretty the same:

```c
#ifdef __WIN32__
    int fd;
    if (ERTS_IS_ATOM_STR("stdout", BIF_ARG_1)) {
        fd = _fileno(stdout);
    } else if (ERTS_IS_ATOM_STR("stderr", BIF_ARG_1)) {
        fd = _fileno(stderr);
    }
#else
// ...
#endif

#ifdef __WIN32__
        written = 0;
        do {
            res = _write(fd, str+written, len-written);
            if (res < 0 && errno != ERRNO_BLOCK && errno != EINTR) {
                printf("[debug] _write went wrong, errno=%d\r\n", errno);
                goto error;
            }
            written += res;
        } while (written < len);
#else
// ...
#endif
```

The test outputted:

```
Testing tests.emulator_test.bif_SUITE.display_string: Starting test, 1 test cases
"hej"
[debug] we're in display_string_2 win32
[debug] and the BIF_ARG_1 is stderr, fd=96
[debug] _write went wrong, errno=9
```

Yet this function seems to be fine when using `erl` either in PowerShell or WSL:

```
$ /mnt/c/src/release/aarch64-windows-msvc/bin/erl.exe
Erlang/OTP 27 [RELEASE CANDIDATE 1] [erts-14.2.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Eshell V14.2.2 (press Ctrl+G to abort, type help(). for help)
1> erlang:display_string("hej").
hejtrue
```

I wonder if we should skip this test for ARM64 Windows or using some alternatives?

cocoa-xu commented:
> I wonder if we should skip this test for ARM64 Windows or using some alternatives?

And for this particular function, `erlang:system_info(system_architecture)` only returns `win32` on all windows systems. Should we consider to change it to `aarch64-pc-windows` by changing these lines of the `echo_env_wsl` function in `otp_build`?

```shell
    if [ X"$X64" = X"true" ]; then
        echo_setenv OVERRIDE_TARGET win32 ';'
        echo_setenv CONFIG_SUBTYPE win64 ';'
    elif [ X"$ARM64" = X"arm64" -o X"$ARM64" = X"x64_arm64" ]; then
        echo_setenv OVERRIDE_TARGET aarch64-pc-windows ';'
        echo_setenv CONFIG_SUBTYPE "$ARM64" ';'
	else
        echo_setenv OVERRIDE_TARGET win32 ';'
    fi
```

dgud commented:
Isn't  *win32* used as the name of the used API and we still use that on arm for windows?

cocoa-xu commented:
> Isn't _win32_ used as the name of the used API and we still use that on arm for windows?

Oh it's probably just a minor thing. I was trying to find a way to detect the CPU architecture using 

```erlang
erlang:system_info(system_architecture).
```

For other systems, the result can usually be used to check the CPU architecture, for example, `x86_64-pc-linux-gnu` or `aarch64-apple-darwin23.2.0`. But for Windows we have to do another check using 

```erlang
os:getenv("PROCESSOR_ARCHITECTURE").
```

Besides that, it also doesn't follow the [official document](https://www.erlang.org/doc/man/erlang.html#system_info_system_architecture):

```
system_architecture
  Returns a string containing the processor and OS architecture the emulator is built for.
```

Maybe it's too late and would definitely break a lot of things for x86 and x86_64 Windows, but perhaps we can do it for ARM64 Windows. WDYT?

cocoa-xu commented:
I still can't figure out why Windows think the file handle we get from `GetStdHandle` is invalid, but I found a workaround, https://github.com/erlang/otp/pull/8142/commits/2d79abba857f885ffd0096f8f31f3ed2e3023ffb, and it fixed all failed test cases due to `erlang:display_string/2`, including:

### emulator
- `bif_SUITE/display_string`
- `hello_SUITE/hello`

### kernel
- `logger_disk_log_h_SUITE/mem_kill_new`
- `logger_disk_log_h_SUITE/restart_after`

I'll upload a screenshot for the `emulator` and `kernel` tests shortly.

cocoa-xu commented:
Some explanation for commit https://github.com/erlang/otp/pull/8142/commits/892019ab1e6fd03381d8f08c2b3642329a7d4130

During my tests, I found that `high` in `erts_time_unit_conversion` can overflow and cause the test case `time_unit_conversion` in `time_SUITE` to fail. For example, on my ARM64 Windows machine when initialising `erts_time_sup__.r.o.start_offset.nsec` in `erts_init_time_sup`, 

```c
    erts_time_sup__.r.o.start_offset.nsec = (ErtsMonotonicTime)
	erts_time_unit_conversion((Uint64) abs_native_offset,
				  (Uint32) ERTS_MONOTONIC_TIME_UNIT,
				  (Uint32) 1000*1000*1000);
```

The value of `ERTS_MONOTONIC_TIME_UNIT` on my machine is `24000000`, and `abs_native_offset` is `576460752312000000`. And `high` will overflow in `erts_time_unit_conversion`, as the minimal example shown below:

```c
#include <stdio.h>
#include <stdint.h>

uint64_t
erts_time_unit_conversion(uint64_t value,
	uint32_t from_time_unit,
	uint32_t to_time_unit)
{
	uint64_t high, low, result;
	if (value <= ~((uint64_t) 0)/to_time_unit)
		return (value*to_time_unit)/from_time_unit;

	low = value & ((uint64_t) 0xffffffff);
	high = (value >> 32) & ((uint64_t) 0xffffffff);
	
	low *= to_time_unit;
	high *= to_time_unit;
	
	high += (low >> 32) & ((uint64_t) 0xffffffff);
	low &= ((uint64_t) 0xffffffff);
	
	result = high % from_time_unit;
	high /= from_time_unit;
	printf("[!] high=%llu\n", high);
	printf("[!] high will overflow=%s\n",  (high >> 32) != 0 ? "yes" : "no");
	high <<= 32;
	printf("[!] high=%llu, result=%llu\n", high, result);
	
	result <<= 32;
	result += low;
	result /= from_time_unit;
	result += high;
	
	return result;
}

int main(int argc, char *argv[]) {
	printf("%lld\n\n", erts_time_unit_conversion(576460752312000000, 24000000, 1000*1000*1000));
	printf("%lld\n", erts_time_unit_conversion(576460752312000000, 24000000, 1000*1000));
}
```

The output of the above minimal code is

```
[!] high=5592405333
[!] high will overflow=yes
[!] high=5572453937501437952, result=9996874
5572453939290448384

[!] high=5592405
[!] high will overflow=no
[!] high=24019196580986880, result=8001996
24019198013000000
```

The correct value of the final result should `24019198012999999488` yet it's too large even for `uint64_t` to hold it. So practically, there's no easy way to get around with it without changing the type of `ErtsMonotonicTime`.

And this overflow will cause all subsequent calculations depending on `ERTS_MONOTONIC_OFFSET_NSEC` to give wrong values. One particular case is that when the user wants to use nano seconds as the time unit in `erlang:monotonic_time/1`.

As for the fix, I think one possible (and perhaps the easiest and the most striaghtforwad) way is that we can simply do the calculations in erlang code. I tried to detect if there's any overflow in the C code and if so, we handle it in erlang code. But after I've done that, I feel that it just adds some unnecessary complexity to the C code.

I hope all this would make sense, but I'm happy to make any other changes to the C code if there's a better solution for it.

cocoa-xu commented:
I was wondering if I should try to use `__int128_t` for `ErtsMonotonicTime` (if `__int128_t` is available) or just leave it for now?

cocoa-xu commented:
Hi all! As this is kinda stale, I wonder if it would be easier for us if I break this PR into smaller PRs? I can start with PRs that definitely won't break any existing things.

dgud commented:
I'm closing this one, since you have started with smaller PR's.
